README.md
# use-progress-bar

Hook to use progress bar in React

## Demo

[Open in CodeSandbox](https://codesandbox.io/s/use-progress-bar-demo-pujdw?fontsize=14)

<p align="center">
  <img width="700" align="center" src="https://user-images.githubusercontent.com/2435004/65696544-12f61c00-e082-11e9-93fe-572ef0d131d0.gif" alt="demo"/>
</p>

## Installation

```bash
$ yarn add use-progress-bar
```

## Usage

```js
import React from 'react'
import { useProgressBar } from 'use-progress-bar'

const ProgressBar = ({ isLoading }) => {
  const { status, progress } = useProgressBar(isLoading)

  const style = {
    opacity: status !== 'progress' ? 0 : 1,
    top: 0,
    position: 'absolute',
    left: 0,
    display: 'block',
    background: '#29d',
    width: `${progress}%`,
    height: '4px',
    transition: `width 200ms ease-in, opacity 200ms ease-in 200ms`,
    zIndex: 10,
  }

  return <div style={style} />
}
```

### `useProgressBar()`

```js
useProgressBar(isLoading: boolean, options?: Options): ProgressBar
```

#### ProgressBar

- `status: 'idle' | 'progress' | 'completed'`
- `progress: number`

#### Options

- `duration?: number`
  - Default: `11000` ms
  - Progress time to the max
- `frequency?: number`
  - Default: `200` ms
  - Frequency of progress updates
- `max?: number`
  - Default: `80`
  - Maximum progress until completion


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.0.0](https://github.com/trzmaxim/use-progress-bar/compare/v1.0.0...v2.0.0) (2020-03-24)

## 1.0.0 (2019-09-26)


### Features

* 🎸 add standard-version ([1fcabcf](https://github.com/trzmaxim/use-progress-bar/commit/1fcabcf))


src/useProgressBar.test.js
// @flow
import { reducer } from './useProgressBar'

const initialState = { status: 'idle', x: 0, progress: 0 }
test('unknown action error', () => {
  expect(() => {
    // $FlowIgnore
    reducer(initialState, { type: 'unknown' })
  }).toThrow('Unknown action')
})

test('progress check', () => {
  let state = initialState
  let prevProgress = 0

  for (let i = 1; i <= 5; i++) {
    state = reducer(state, { type: 'increment', increment: 1, max: 100 })
    expect(state.status).toBe('progress')
    expect(state.progress > prevProgress).toBe(true)
    prevProgress = state.progress
  }

  state = reducer(state, { type: 'complete' })
  expect(state.status).toBe('completed')
  expect(state.progress).toBe(100)

  state = reducer(state, { type: 'reset' })
  expect(state).toEqual(initialState)
})


src/useProgressBar.test.ts
import { reducer, State } from './useProgressBar'

const initialState: State = { status: 'idle', x: 0, progress: 0 }
test('unknown action error', () => {
  expect(() => {
    // @ts-ignore
    reducer(initialState, { type: 'unknown' })
  }).toThrow('Unknown action')
})

test('progress check', () => {
  let state = initialState
  let prevProgress = 0

  for (let i = 1; i <= 5; i++) {
    state = reducer(state, { type: 'increment', increment: 1, max: 100 })
    expect(state.status).toBe('progress')
    expect(state.progress > prevProgress).toBe(true)
    prevProgress = state.progress
  }

  state = reducer(state, { type: 'complete' })
  expect(state.status).toBe('completed')
  expect(state.progress).toBe(100)

  state = reducer(state, { type: 'reset' })
  expect(state).toEqual(initialState)
})


src/useProgressBar.js
// @flow
import * as React from 'react'

import type { ProgressBar, Options, State, Action } from './types.js.flow'

const IDLE = 'idle'
const PROGRESS = 'progress'
const COMPLETED = 'completed'

const MAX_X = 60

const DEFAULT_FREQUENCY = 200 // ms
const DEFAULT_MAX = 80
const DEFAULT_DURATION = 11000 // 11 sec

const initialState: State = { status: IDLE, x: 0, progress: 0 }
export const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case 'increment': {
      const { increment, max } = action
      const x = state.x + increment
      const progress =
        Math.round((1 - Math.exp((-1 * x) / 10)) * 100 * max) / 100

      return {
        status: PROGRESS,
        x,
        progress: Math.max(progress, state.progress),
      }
    }
    case 'complete':
      return {
        ...state,
        status: COMPLETED,
        progress: 100,
      }
    case 'reset':
      return initialState
    default:
      throw new Error('Unknown action')
  }
}

const useProgressBar = (
  isLoading: boolean,
  {
    frequency = DEFAULT_FREQUENCY,
    max = DEFAULT_MAX,
    duration = DEFAULT_DURATION,
  }: Options = {},
): ProgressBar => {
  const [{ x, progress, status }, dispatch] = React.useReducer<State, Action>(
    reducer,
    initialState,
  )

  React.useEffect(
    () => dispatch(!isLoading ? { type: 'complete' } : { type: 'reset' }),
    [isLoading],
  )

  React.useEffect(() => {
    if (!isLoading || x > MAX_X) {
      return
    }

    const increment = MAX_X / (duration / frequency)
    const id = setTimeout(
      () => dispatch({ type: 'increment', increment, max }),
      frequency,
    )

    return () => clearTimeout(id)
  }, [isLoading, x])

  return { status, progress }
}

export default useProgressBar


src/useProgressBar.ts
import React from 'react'

export interface ProgressBar {
  readonly status: 'idle' | 'progress' | 'completed'
  readonly progress: number
}

export interface Options {
  readonly frequency?: number
  readonly max?: number
  readonly duration?: number
}

export interface State {
  readonly status: 'idle' | 'progress' | 'completed'
  readonly x: number
  readonly progress: number
}

export type Action =
  | { type: 'increment'; increment: number; max: number }
  | { type: 'complete' }
  | { type: 'reset' }

const IDLE = 'idle'
const PROGRESS = 'progress'
const COMPLETED = 'completed'

const MAX_X = 60

const DEFAULT_FREQUENCY = 200 // ms
const DEFAULT_MAX = 80
const DEFAULT_DURATION = 11000 // 11 sec

const initialState: State = { status: IDLE, x: 0, progress: 0 }
export const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case 'increment': {
      const { increment, max } = action
      const x = state.x + increment
      const progress =
        Math.round((1 - Math.exp((-1 * x) / 10)) * 100 * max) / 100

      return {
        status: PROGRESS,
        x,
        progress: Math.max(progress, state.progress),
      }
    }
    case 'complete':
      return {
        ...state,
        status: COMPLETED,
        progress: 100,
      }
    case 'reset':
      return initialState
    default:
      throw new Error('Unknown action')
  }
}

export const useProgressBar = (
  isLoading: boolean,
  {
    frequency = DEFAULT_FREQUENCY,
    max = DEFAULT_MAX,
    duration = DEFAULT_DURATION,
  }: Options = {},
): ProgressBar => {
  const [{ x, progress, status }, dispatch] = React.useReducer(
    reducer,
    initialState,
  )

  React.useEffect(
    () => dispatch(!isLoading ? { type: 'complete' } : { type: 'reset' }),
    [isLoading],
  )

  React.useEffect(() => {
    if (!isLoading || x > MAX_X) {
      return
    }

    const increment = MAX_X / (duration / frequency)
    const id = setTimeout(
      () => dispatch({ type: 'increment', increment, max }),
      frequency,
    )

    return () => clearTimeout(id)
  }, [isLoading, x])

  return { status, progress }
}
